### PR TITLE
Upgraded to MathJax v4, caching, and changed inline alignment method.

### PR DESCRIPTION
**Pull request by coreyb-git**
This is an overhaul.

Added:

- Upgraded to MathJax v4.
- SVG caching for much faster rebuilds of ADoc files (smarter caching than the pull request for v2).
- Inline alignment resolved in a simpler way by enforcing a normalized height across all inline SVG's, using \vphantom text, and removing extra padding added by MathJax for oversized SVG's such as \vec A (the arrow creates a small amount of extra padding that normally produces a taller SVG with a vertical center that is offset from surrounding text.  From what I've seen thus far this aligns the text.
- Have made the test suite conform to this starting state of the V4+cache version.
- A lot of debug logging in verbose mode.

Added attributes:
- :math-cache-dir:
- :math-debug:
- :math-debug-color:
- :math-inline-body-scale:
- :math-inline-heading-scale:

The scale values allow for tweaking for those ADoc theme fonts that are slightly different in stroke thickness and overall height.  From brief testing 0.8 is the perfect value to tweak the inline body and heading math text to re-align everything when the theme text is slightly smaller.

Removed:

- Currently unable to select different math fonts.  Probably a todo item for later.
- Removed V2 test pdf files

Issues:
- Block LaTeX can overflow into the footer and off the bottom of the page making typesetting a different challenge than that of asciidoctor-mathematical.
- Due to the way I've handled unaligned inline text the descenders can be chopped for inline text if they are extra deep (like with subscripts on integrals).  I consider this a satisfactory sacrifice for the sake of an easier alignment scheme, as it's reasonable to NOT expect complex/tall LaTeX math within inline text.  To get an intuitive idea of what is happening the verification pdf's with the SVG backgrounds colored in beige, and the phantom chars visible, demonstrates how I've handled aligning the text.

TODO:
- Investigate if there's a way to enforce the block LaTeX as unbreakable nodes.
- Add support for changing math font (user has to manually install each specific extra font packages, as per MathJax V4 documentation.  Fonts don't all come-with anymore due to overall size.
- Update the README.md

Probably other things I've added/removed but have forgotten.  This has consumed more time than I ever intended.  Again, I'm not a ruby dev... or wasn't before this.  This overhaul is borne from my need of better typesetting than was provided by asciidoctor-mathematical and the original asciidoctor-pdf-mathjax.

### CODE REVIEW

**Comment by Crown0815** [comment]
Test are failing

**Comment by Crown0815** [comment]
@coreyb-git the changes look great.
Currently the output contains the _normalization_ symbols that you added. Those would need to be removed before merging.

**Comment by coreyb-git** [comment]
That should just be due to the debug attribute being set in the testing config file. Under normal operation they should not appear.  I'll have more commits soon...

I'm currently doing a little more work on the alignment code to see if I can make it slightly more robust.

**Comment by coreyb-git** [comment]
If you meant specifically within the block math?  Found that bug.

**Comment by coreyb-git** [comment]
Hi @Crown0815, 

I think I have hit a reasonable pause point with functionality.

- I've changed the normalization prefix for a better alignment.  
- Fixed the bug you mentioned with the block LaTeX retaining debug renders.  
- I've processed new, clean, verification pdf's, except the custom font test pdf.  There is an issue with Level 0 headings being pushed downwards by the tall SVG's that needs addressing before creating a verification pdf with the newcm math font.
- I've also added to the workflow YAML the instructions to install node and mathjax-full... but this part was pure vibe-coded with Gemini, so you'll need to review that and confirm that it works or not.  This is a continuation from removing the setup and teardown code in the test suite.
- Added a default scaling of 1.2 for block stem nodes.  Configurable with :math-body-scale:
- Debug levels are currently 0 (default), 1 adds just the background and also a baseline line, and 2 adds that plus the normalization prefix instead of hiding it as a vphantom.

Known Issues:
- Level 0 headings get pushed down from the top by the SVG image (too tall).
- Block stem can encroach into book footer.  Asciidoctor-mathematical doesn't do this, so there must be another way.
- README hasn't been updated... low priority over getting this working well.
- Font customization hasn't been addressed at all.

I think that's it.  If you get time hit me with any feedback/thoughts.

File: .github/workflows/build-and-release.yaml [hunk 1]
```diff
@@ -19,50 +17,47 @@ jobs:
...
     runs-on: ubuntu-latest
     steps:
       - name: Checkout code
         uses: actions/checkout@v4
-
       - name: Set up Ruby
         uses: ruby/setup-ruby@v1
-
+        # Add this under the 'test' job, right after 'Set up Ruby'
+      - name: Set up Node
```
Comment (Crown0815) [inline .github/workflows/build-and-release.yaml:52]: This is not working and should not be necessary. Since node is already preinstalled on the ubuntu latest runner.
The installation of mathjax is supposed to be done in the test setup to keep things simple

File: .gitignore [hunk 2]
```diff
@@ -14,3 +14,6 @@ Gemfile.lock
 node_modules/
 package-lock.json
 package.json
+
+#cache dir
+test/mathcache/
```
Comment (Crown0815) [inline .gitignore:19]: Better to ignore the filetype than the directory. SVG I assume?

File: README.md [hunk 3]
```diff
@@ -41,6 +116,9 @@ You can find an example of a docker container configuration for Asciidoctor PDF
 
 ### Installation from RubyGems
 
+> [!WARNING]
```
Comment (Crown0815) [inline README.md:119]: Please remove this warning

File: README.md [hunk 4]
```diff
@@ -72,6 +150,10 @@ Here’s how to convert an AsciiDoc file with mathematical content to PDF using
    ```asciidoc
    = Document with Math
    :stem:
+   :math-cache-dir: <Your Cache Dir>
+   :math-font: <Your Chosen MathJax v2 Font>
```
Comment (Crown0815) [inline README.md:154]: This is not correct currently, right?
What is currently the issue with supporting the fonts?
MathJax 4 seems to support many different fonts (https://docs.mathjax.org/en/v4.0/output/fonts.html).

File: README.md [hunk 5]
```diff
@@ -72,6 +150,10 @@ Here’s how to convert an AsciiDoc file with mathematical content to PDF using
...
+
+   //End of heading and attribute settings.
```
Comment (Crown0815) [inline README.md:156]: This comment is not useful

File: README.md [hunk 6]
```diff
@@ -81,11 +163,39 @@ Here’s how to convert an AsciiDoc file with mathematical content to PDF using
...
 
-2. Run the `asciidoctor-pdf` command with the extension:
+> [!IMPORTANT]
+> Don't forget to set the cache directory attribute to a location on your HDD where you want the cached SVG image files to live.  Without this attribute set the SVG images are regenerated each time you build your PDF, and then destroyed once the SVG data has been embedded into the PDF.
+
+> [!TIP]
+> You don't have to specify the math-font attribute if you are satisfied with the default font used by MathJax.
+
+> [!TIP]
+> Be aware that the first blank line in an ADoc file signals the end of the header/settings.  Ensure any attributes you wish to
```
Comment (Crown0815) [inline README.md:173]: This is general asciidoc information, not appropriate in here

File: README.md [hunk 7]
```diff
@@ -81,11 +163,39 @@ Here’s how to convert an AsciiDoc file with mathematical content to PDF using
    ++++
    ```
 
-2. Run the `asciidoctor-pdf` command with the extension:
+> [!IMPORTANT]
+> Don't forget to set the cache directory attribute to a location on your HDD where you want the cached SVG image files to live.  Without this attribute set the SVG images are regenerated each time you build your PDF, and then destroyed once the SVG data has been embedded into the PDF.
```
Comment (Crown0815) [inline README.md:167]: This makes it sound like the cache is mandatory. It should be the other way around.
If your conversion is slow, you can speed it up by using the cache

File: README.md [hunk 8]
```diff
@@ -81,11 +163,39 @@ Here’s how to convert an AsciiDoc file with mathematical content to PDF using
...
    ```shell
-   asciidoctor-pdf -r asciidoctor-pdf-mathjax mathdoc.adoc -o mathdoc.pdf
+   asciidoctor-pdf \
+      -r asciidoctor-pdf-mathjax \
+      mathdoc.adoc \
+      -o mathdoc.pdf
    ```
-   - `-r asciidoctor-pdf-mathjax`: Loads the extension.
+
+   - Load the extension, depending on where you got it from:
```
Comment (Crown0815) [inline README.md:194]: This seems like a duplication of the lines above

File: README.md [hunk 9]
```diff
@@ -94,22 +204,43 @@ Here’s how to convert an AsciiDoc file with mathematical content to PDF using
...
+  failures during the rendering process. Issues with specific stem blocks, that aren't
+  outright crashes of this extension, indicate a peculiarity with the LaTeX notation you have used that is preventing it
+  from being understood, and thus it can't be rendered as an SVG image and embedded into the final PDF.  Check your notation
+  for any minor issues you may have overlooked.
+
+#### Supported Fonts
+
+By setting the `:math-font: <Font Name>` attribute in the heading of your adoc file, or with `-a math-font=<Font Name>` on the command line, you can choose the font you prefer.  Available fonts in MathJax v2 are:
+
+- TeX: The default Computer Modern look (like standard LaTeX).
```
Comment (Crown0815) [inline README.md:223]: These are not correct. Again, any chance to bring in support for https://docs.mathjax.org/en/v4.0/output/fonts.html?

File: README.md [hunk 10]
```diff
@@ -122,6 +253,10 @@ To set up your development environment to run the test you need to install the f
 - [Ruby](https://www.ruby-lang.org/) 3.4 or higher
 - [NodeJS](https://nodejs.org/en)
 - [diff-pdf](https://github.com/vslavik/diff-pdf)
+- [ruby-minitest](https://github.com/minitest/minitest)
+
+> [!TIP]
+> in the `/test` directory is the script `run_tests.sh` to automate testing.  It creates a Docker/Podman image with diff-pdf and ruby-minitest installed, and then it runs the tests in the `/test` directory.
```
Comment (Crown0815) [inline README.md:259]: Please remove this tip. The testing inside docker may be great, but it also creates an additional hurdle for anybody who already has ruby and node, who now also has to install docker

File: test/run_podman_tests.sh [hunk 11]
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
+podman build --tag run_ruby_tests_image .
```
Comment (Crown0815) [inline test/run_podman_tests.sh:2]: This is a very specific test script.
I believe the tests should be run in ruby directly.

Please keep this script out of the PR.

File: test/test_asciidoctor_pdf_mathjax.rb [hunk 12]
```diff
@@ -9,14 +9,16 @@
 
 class TestAsciidoctorPdfMathjax < Minitest::Test
   def setup
-    system("npm install --silent mathjax-node")
+    # system("npm install --silent mathjax-node")
```
Comment (Crown0815) [inline test/test_asciidoctor_pdf_mathjax.rb:12]: The tests were purposely designed install mathjax-node.
Please upgrade it to mathjax-full and leave the removal in as well.

File: test/test_asciidoctor_pdf_mathjax.rb [hunk 13]
```diff
@@ -29,22 +31,26 @@ def teardown
...
 
   def test_that_conversion_with_custom_font_theme_works
-    verify_conversion_of('stem_latex', {'pdf-theme' => 'custom-font'})
+    verify_conversion_of('stem_latex', { 'pdf-theme' => 'custom-font' })
   end
 
-
-  MATHJAX_FONTS = %w[TeX STIX-Web Asana-Math Neo-Euler Gyre-Pagella Gyre-Termes Latin-Modern]
+  # MATHJAX_FONTS = %w[TeX STIX-Web Asana-Math Neo-Euler Gyre-Pagella Gyre-Termes Latin-Modern]
+  MATHJAX_FONTS = %w[mathjax-newcm]
```
Comment (Crown0815) [inline test/test_asciidoctor_pdf_mathjax.rb:38]: What happens if you put in the other fonts?

**Review by Crown0815 — CHANGES_REQUESTED** [review]
In the verification tests I see that the math font is not aligned based on the x-height anymore. Is is now aligned based on the M-height?

I am a little worried about this update, since it changes the alignment and therefore would like to understand the goal.

I assume using the M-height reduces the amount of "too large" formulas being cropped?

**Comment by coreyb-git** [comment]
I've lost too much time to refactoring this plugin and have to put it aside for now, so I'll close this pull request.  I'll address a few things, for context.

- If I get time later in the year I might come back to this and pick it back up.
- I am writing a large math-heavy document, so I put time into this because I actively need it right now (especially caching). 
- The pull request was so that you could, if you wanted, take what I've done and run further with it.   Since it's robust enough for my use my focus has shifted back to my writing, out of necessity.
- x-height and m-height... The goal of the alignment logic right now is to not fight against how Prawn inserts and aligns inline images.  I've made an extensive note in the ruby file itself at the top, but in short I'm intentionally pushing the height of the image out in both directions such that it reaches (or as close as possible) the lower bound that exists, which is somewhere at or near the bottom of the descender.  This changes depending on the font, so the vphantom text can need to be customized when not using the default Noto Serif asciidoctor-pdf font, which is a feature I've added.  Using vphantom characters removes the need to try and rework the svg heights and viewbox data, and prevents the decapitation of LaTeX text if the math is wrong.   It's a one-and-done hacky fix per font.
- Some of the README changes were for the other pull request, before I brought-in MathJax v4 and reworked everything, so the readme is entirely wrong, and hasn't been a priority after the update.
- The font issue with v4, as I currently understand it (right or wrong) is that the supported fonts are NOT included with the mathjax-full install, and anything other than the default needs to be manually installed by the user.  This isn't a major issue, but I simply haven't gotten to the point of updating the render.js file to support anything other than the default, yet, regardless of what fonts are or are not installed.  Not a difficult addition, just not something I got to.  There's also a large chance I misread or conflated documentation somewhere and adding support in render.js for these other fonts may be much simpler than I had understood things to be, though separate installation of each font does seem mandatory to have anything other than newcm.

In short, you are welcome to take what I've done so far and continue with it, if any part of it is of interest to you (if you don't like the alignment logic there's still the caching feature, for example).  For now, I have to shift focus elsewhere, and may come back to it later.

Thanks for creating this extension.  It works so much better than asciidoctor-mathematical, and I've been able to craft it into a tool that is stronger again specific to my requirements.

**Comment by Crown0815** [comment]
@coreyb-git, thank you for the clarification. I will see that your changes will be integrated.
Thanks a lot for the hard work you put into this. I can see the effort and am deeply grateful for it.

Best of luck with your writing!
